feat: Wwise documentation (#359)
satisfactorymodding/Documentation: modules/ROOT/pages/Development/Satisfactory/Audio.adoc

The parameter must have a range defined.
A useful range is `0.0` to `1.0` since many existing gameand engine systems already use that range as a start-finish, minimum-maximum, or off-on scale.

Now, assign the RTPC Game Parameter to your audio.

Open your Audio in Wwise and go to the `RTPC` tab.
Click the `>>` in the main section to set what you want to update.
This example case shows updating the `Voice Volume`.

Then click the `>>` to the right of the first one and select your game parameter.

image:Satisfactory/Wwise/Wwise_RTPC_Configure.png[Wwise_RTPC_Configure, 500]

Once that is done, you should see something similar to the following:

image:Satisfactory/Wwise/Wwise_RTPC_Overview.png[Wwise_RTPC_Overview, 100%]

Now, to create the link between Unreal Engine and Wwise you will need to create the Game Parameter in Unreal Engine:

image:Satisfactory/Wwise/Wwise_UE_RTPC_Create.png[Wwise_UE_RTPC_Create, 600]

Name it the same as you have in Wwise.

You can now update the RTPC Game Parameter value using blueprints:

image:Satisfactory/Wwise/Wwise_UE_RTPC_Usage.png[Wwise_UE_RTPC_Usage, 700]

=== Playback Limits

Wwise limits the amount of audio that can be played at once. 

> Trying to play more than this limit will result in lower-priority sounds like footsteps, music, or animal sounds playing.

Reviewer: Did this get flipped? I would expect higher-priority sounds to be played in this situation, not the lower priority ones.

This means that when the limit for this audio is reached, it will kill any oldest instances and discard them.

Finally, you will need to set up playback priority. The playback priority requires that you set up Attenuation.

Set the priority of your audio to something like `10`, then set the `Offset priority by` setting to `-10`.

This setting will set the audio priority to 10, and then when you are outside the attenuation distance, it will set the priority to 0, allowing for more important audio, like the game sounds, to be played.

Even with Attenuation and playback limits set, the audio will still play and count towards the Wwise Playback Limit.

* Attenuation will only lower the volume of the audio and not stop the audio.
* Playback limits only limit the number of audio instances playing on a game object.

So, to prevent audio from playing when you are outside of the attenuation range, you will need to use the Significance Manager.

For playing background building sounds, use the `Gained Significance` and `Lost Significance` events:

image:Satisfactory/Wwise/Wwise_Significance.png[Wwise_Significance, 700]

> For fire and forget ak events, use the `Get Is Significant` function before posting the event:

Reviewer: ```suggestion
For fire and forget AK events, use the `Get Is Significant` function before posting the event:
```

== Wwise

Coffee Stain uses the Wwise audio system to play sounds in Satisfactory.

To create new audio events to play in your mod, you will need the Wwise project set up and integrated into the SML project.

=== Setting Up Audio Buses

By default when using Wwise audio only the `Master` volume slider will work.

You will need to manually create the different Audio Buses that link to the other sliders.

TODO: Waiting for CSS to provide the correct audio bus names.

=== Importing Media Files to Wwise

Open the Wwise project file, typically in `SMLProject/SMLProject_WwiseProject/SMLProject_WwiseProject.wproj`. Double-clicking this file will open Wwise.

To make sure you are in the correct layout at the top, go to `Layout -> Designer`

On the left-hand side, you will see the project explorer; select the Audio Tab.

> image:Satisfactory/Wwise/Wwise_Image1.png[Wwise_Image1, 350]

Reviewer: Could you rename the image files to talk about what they are showcasing? Numbered file names is troublesome as inevitably another one will be added in the middle of the sequence

Author: Renamed images for better understanding.

==== Fire & Forget Events

If you would like to play a one-off AK event, such as a spark sound, you will need to use the following method in your blueprint.

image:Satisfactory/Wwise/Wwise_Image13.png[Wwise_Image13, 100%]

This will first create a new `AK Component` at the blueprint's root component and then play the AK Event specified.

Once the AK event has ended, you must clean up the `AK Component` using the callback event.

==== Looping Events

If you would like to play a sound effect that might loop, for example, a humming/buzzing sound for a building.

The best method is to add an `AK Component` to your blueprint and set the `AK Audio Event` variable on the component.

You must ensure that you stop the event when the actor is destroyed.

image:Satisfactory/Wwise/Wwise_Image14.png[Wwise_Image14, 100%]

=== Setting Up Attenuation

If you plan to use Wwise audio for buildings, you will need to set up attenuation. By default, the audio will play even if you are on the other side of the map.

Wwise has a great tutorial on how to set this up here: https://youtu.be/lOP0seni1DY[Wwise & Unreal | Adding 3D Audio With Attenuation]

=== Updating Wwise Values From Your Mod

> You can update audio values using RTPC (Game Parameters).

Reviewer: what does RTPC stand for? it sounds like Game Parameters is part of the Unreal asset side of things

Author: Updated with better explaination on what RTPC means and clearer understanding of what it does.

image:Satisfactory/Wwise/Wwise_Image15.png[Wwise_Image15, 100%]

Name this whatever you want.

Open the newly created Game Parameter, and you should see the following in the main section.

image:Satisfactory/Wwise/Wwise_Image17.png[Wwise_Image17, 100%]

You will need to set the range. I recommend using 0 to 1.

Now, assign the RTPC Game Parameter to your audio.

Open your Audio and go to the `RTPC` tab. Click the `>>` in the main section to set what you want to update. In this case, I want to update the `Voice Volume`.

Then click the `>>` to the right of the first one and select your game parameter.

image:Satisfactory/Wwise/Wwise_Image16.png[Wwise_Image16, 500]

Once that is done, you should see something similar to the following:

image:Satisfactory/Wwise/Wwise_Image18.png[Wwise_Image18, 100%]

Now, create the Game Parameter in Unreal Engine.

image:Satisfactory/Wwise/Wwise_Image19.png[Wwise_Image19, 500]

Name it the same as you have in Wwise.

You can now update the RTPC Game Parameter value using blueprints:

image:Satisfactory/Wwise/Wwise_Image20.png[Wwise_Image20, 700]

> 

Reviewer: What considerations need to be made for source control? I think the wwise project is currently excluded from git, and that will need to be changed on the SML side eventually. What is the RP team doing?

Author: We are currently adding the Wwise project to a separate repo. This contains all the source `.wav` files and all the settings for our audio. We are using Git LFS to store the larger files like the `.wav` files as per Wwise recommendations here: 

[Using Wwise with your source control system](https://www.audiokinetic.com/en/library/edge/?source=Help&id=using_wwise_with_source_control_system)

I recommend that SML create a new repo for the Wwise project that contains the Stubs and other info from the CSS audio team. Then, as part of the SML toolkit setup, this has to be pulled into the project. Then, if the mod dev uses Wwise, they can just change the git remote to their repo URL.

Reviewer: ```suggestion
== Wwise and Source Control

[NOTE]
====
This heading is a work in progress.
The SML team plans to ship a Wwise project that includes stubs for base game Wwise events in the future.
The suggested approach in this heading may change once that process is complete.
====

Wwise project files require extra consideration to store in
xref:Development/BeginnersGuide/LearnGit.adoc[source control].
Distributing the raw audio assets is sometimes prohibited by audio licenses.

One approach is to create a separate (private) repository to store the Wwise project and to enable Git LFS for raw audio files.
However, this interferes with the typical one-repo-per-mod structure by tying multiple mod's files together in one repository.

Wwise offers advice on connecting to source control
https://www.audiokinetic.com/en/library/edge/?source=Help&id=using_wwise_with_source_control_system[in their documentation].
```

== Wwise

Coffee Stain uses the Wwise audio system to play sounds in Satisfactory.

To create new audio events to play in your mod, you will need the Wwise project set up and integrated into the SML project.

=== Setting Up Audio Buses

By default when using Wwise audio only the `Master` volume slider will work.

You will need to manually create the different Audio Buses that link to the other sliders.

> TODO: Waiting for CSS to provide the correct audio bus names.

Reviewer: What should people do in the mean time? Do we currently know any of the audio bus names? These, for example? https://discord.com/channels/555424930502541343/601030071221878784/1294145591709794379

Author: ATM, the master audio bus that Wwise has by default, correctly links to the in-game Master volume slider. This is currently the only way to change the volume of Wwise audio.

The variables linked to that Discord message are RTPC variables and not the names of the Audio Buses. I have tried a few different combinations of those just incase it worked. But this did not work. We will need to wait for CSS to provide the names of the Audio Buses.

Go To `Audiokinetic -> Audiokinetic Event`

image:Satisfactory/Wwise/Wwise_UEAkEvent_Create.png[Wwise_UEAkEvent_Create, 500]

Name the new asset **Exactly** the same name as the event in Wwise.

[NOTE]
====
If the asset name is exactly the same as the Wwise event, Unreal Engine will automatically populate the correct event information to the variables below.
====

Verify that the event has the correct information by opening the event in Unreal.

* Verify that the `Wwise Short ID` matches the event ID in Wwise.
* Verify the `Wwise Name` matches the event name.

image:Satisfactory/Wwise/Wwise_UEAkEvent_Verify.png[Wwise_UEAkEvent_Verify, 500]

If everything has worked correctly, you can now test the event by right-clicking the event in Unreal and clicking `Play Event`

=== Playing AKEvents In Blueprints

==== Fire & Forget Events

If you would like to play a one-off AK event, such as a spark sound, you will need to use the following method in your blueprint.

image:Satisfactory/Wwise/Wwise_UEAkEvent_FireAndForget.png[Wwise_UEAkEvent_FireAndForget, 100%]

This will first create a new `AK Component` at the blueprint's root component and then play the AK Event specified.

> Once the AK event has ended, you must clean up the `AK Component` using the callback event.

Author: @budak7273 Correct. Technically, one AK component can trigger multiple AK events simultaneously.

So, if you want to trigger multiple events or just occasionally trigger an event, it's best to add an AK Component to the actor.

> Once the AK event has ended, you must clean up the `AK Component` using the callback event.

Author: Speaking with Arch, the AK component for both fire and forget and AK component attached to the component will be destroyed once the actor is destroyed.

==== Looping Events

If you would like to play a sound effect that might loop, for example, a humming/buzzing sound for a building.

The best method is to add an `AK Component` to your blueprint and set the `AK Audio Event` variable on the component.

You must ensure that you stop the event when the actor is destroyed.

image:Satisfactory/Wwise/Wwise_UEAkEvent_EndPlay.png[Wwise_UEAkEvent_EndPlay, 100%]

> === Setting Up Attenuation

Author: Yeah, stuff like UI and ADA messages won't require attenuation. I'm going to add steps to set this up, as the video feels a bit of a cop-out.

> === Setting Up Attenuation

Author: Updated the docs to include setting up attenuation

If you plan to use Wwise audio for buildings, you will need to set up attenuation. By default, the audio will play even if you are on the other side of the map.

Wwise has a great tutorial on how to set this up here: https://youtu.be/lOP0seni1DY[Wwise & Unreal | Adding 3D Audio With Attenuation]

=== Updating Wwise Values From Your Mod

> Wwise uses a feature called Real-Time Parameter Control (RTPC) that allows game engines like Unreal or Unity to change values on the Wwise software in real-time.
> 
> An example of this would be if, for example, you have a building and want the sound volume to fluctuate based on the recipe's progress.
> 
> You can set up an RTPC in Wwise and Unreal and update the value in your building's blueprint.

Reviewer: ```suggestion
Wwise uses a feature called Real-Time Parameter Control (RTPC) that allows game engines like Unreal or Unity to change values on Wwise assets in real-time.
This enables special behaviors like a manufacturing building's sound volume to fluctuate based on recipe progress.

After setting up RTPC in Wwise and Unreal, blueprint code can update audio parameter values.
```

To set these up, in your Wwise project, go to Project Explorer and the `Game Syncs` tab.

Under `Game Parameters -> Default Work Unit` create a new game parameter by right-clicking the `Default Work Unit` and then `New Child -> Game Parameter`

image:Satisfactory/Wwise/Wwise_RTPC_Create.png[Wwise_RTPC_Create, 100%]

Name this whatever you want.

> Open the newly created Game Parameter, and you should see the following in the main section.

Reviewer: ```suggestion
Open the newly created Game Parameter, and you should see the following in the main section:
```

image:Satisfactory/Wwise/Wwise_RTPC_Range.png[Wwise_RTPC_Range, 100%]

> You will need to set the range. I recommend using 0 to 1.

Reviewer: ```suggestion
The parameter must have a range defined.
A useful range is `0.0` to `1.0` since many existing gameand engine systems already use that range as a start-finish, minimum-maximum, or off-on scale.
```

Now, assign the RTPC Game Parameter to your audio.

> Open your Audio and go to the `RTPC` tab. Click the `>>` in the main section to set what you want to update. In this case, I want to update the `Voice Volume`.

Reviewer: ```suggestion
Open your Audio in Wwise and go to the `RTPC` tab.
Click the `>>` in the main section to set what you want to update.
This example case shows updating the `Voice Volume`.
```
